### PR TITLE
Pairing & multi device & group-chats

### PR DESCRIPTION
**Pull request by cammellos**
### Pairing

Add a new section (under advanced) where you can pair new devices.
I have included a commit from a different PR as both rely on a status-go update and probably best to be tested together.

The code is behind a feature flag but enabled in nighties if development mode is on.
The UI is not polished and this is only the first step, but I decided to enable it in nightlies in dev-mode as it can already provide value to whoever wants to use it.

It is only used to sync up your devices contacts. Group chats will also receive your own messages once devices are paired correctly.

### What's left

1) Polish UI
2) Let the user manually confirm the devices (currently they are enabled without user confirmation)
3) Add information about each device (is this desktop/mobile/last-seen)
4) Sync other kind of info (your name etc)
5) Keep data in sync after initial pairing (contacts, 1-to-1 messages)

### Testing

Both group chats & pairing should only be enabled in nightlies/dev if dev-mode is on.

To pair your devices:

1) Install this build on both desktop & mobile
2) On android/ios go to settings->advanced->devices, on desktop go to Profile->Devices->Pair
3) Pressing `+` will advertise your device. You should now see a new ID in all the `devices` sections.
4) Do the same on the other devices until all your devices shows up in the list.
5) Press on any device ID to share your contacts with all the rest of the devices.
6) Group chats should receive your own messages now, if you start a new group chat with someone, their message should be received by all the paired devices.

status: ready

### CODE REVIEW

**Review by janherich — APPROVED** [review]

**Comment by cammellos** [comment]
@janherich could you review again as significant changes were made since you approved it? Thanks!

File: src/status_im/pairing/core.cljs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  (:require
+   [status-im.utils.fx :as fx]
+   [status-im.utils.config :as config]
+   [status-im.transport.message.protocol :as protocol]
+   [status-im.data-store.installations :as data-store.installations]
+   [status-im.data-store.contacts :as data-store.contacts]
+   [status-im.transport.message.pairing :as transport.pairing]))
+
+(defn start [cofx]
+
```
Comment (rasom) [inline src/status_im/pairing/core.cljs]: why empty line here?

File: src/status_im/pairing/core.cljs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  (let [{:keys [current-public-key web3]} (:db cofx)]
+    {:shh/send-pairing-message {:web3          web3
```
Comment (rasom) [inline src/status_im/pairing/core.cljs]: values in the map are aligned like there was some longer key

File: src/status_im/pairing/core.cljs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    {:shh/send-direct-message [{:web3 web3
+                                :src     current-public-key
+                                :dst     current-public-key
+                                :payload payload}]}))
+
+(defn handle-sync-installation [{:keys [db] :as cofx} {:keys [contacts]} sender]
+  (let [dev-mode? (get-in db [:account/account :dev-mode?])]
+    (when (config/pairing-enabled? dev-mode?)
+      (when (= sender (get-in cofx [:db :current-public-key]))
```
Comment (rasom) [inline src/status_im/pairing/core.cljs]: can be reduced to one `when` form

File: src/status_im/pairing/core.cljs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+(defn sync-installation-message [{:keys [db]}]
+  (transport.pairing/SyncInstallation. (:contacts/contacts db)))
+
+(defn send-installation-message [cofx]
+  (protocol/send (sync-installation-message cofx) nil cofx))
+
+(defn send-fx [{:keys [db]} payload]
+  (let [{:keys [current-public-key web3]} db]
+
```
Comment (rasom) [inline src/status_im/pairing/core.cljs:51]: extra empty line

File: src/status_im/pairing/core.cljs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    {:shh/send-direct-message [{:web3 web3
```
Comment (rasom) [inline src/status_im/pairing/core.cljs]: alignment in the map

File: src/status_im/pairing/core.cljs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  (let [dev-mode? (get-in db [:account/account :dev-mode?])]
+    (when (config/pairing-enabled? dev-mode?)
+      (when (= sender (get-in cofx [:db :current-public-key]))
+        (let [new-contacts  (merge-contacts (:contacts/contacts db) contacts)]
+          {:db (assoc db :contacts/contacts new-contacts)
+           :data-store/tx [(data-store.contacts/save-contacts-tx (vals new-contacts))]})))))
+
+(fx/defn load-installations [{:keys [db all-installations]}]
+  {:db (assoc db :pairing/installations (reduce (fn [acc {:keys [installation-id] :as i}]
+                                                  (assoc acc installation-id i)) {} all-installations))})
```
Comment (rasom) [inline src/status_im/pairing/core.cljs]: not a big big deal, but can we fit onto 80 chars here?

File: src/status_im/ui/screens/pairing/views.cljs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            [status-im.ui.components.list.views :as list]
+            [status-im.ui.components.react :as react]
+            [status-im.ui.components.status-bar.view :as status-bar]
+            [status-im.ui.components.toolbar.view :as toolbar]
+            [status-im.ui.components.toolbar.actions :as toolbar.actions]
+            [status-im.ui.screens.profile.components.views :as profile.components]
+            [status-im.ui.screens.pairing.styles :as styles]))
+
+(defn synchronize-installation! [id]
+  (re-frame/dispatch [:pairing.ui/synchronize-installation-pressed id]))
```
Comment (dmitryn) [inline src/status_im/ui/screens/pairing/views.cljs:17]: no new line between `defn`

File: src/status_im/pairing/core.cljs (hunk 4 above)
Comment (dmitryn) [inline src/status_im/pairing/core.cljs:51]: empty line

**Review by dmitryn — APPROVED** [review]
lgtm, except minor formatting issues

**Comment by cammellos** [comment]
Screenshots, obviously UI has not been polished/nor worked on, this only adds the mechanism, it would be good to get some feedback on how it should work, thanks!
android
desktop

**Comment by lukaszfryc** [comment]
@cammellos does this PR enable group chats? If yes, could you update its title to include this change?

**Comment by cammellos** [comment]
@lukaszfryc will do thanks, hold on testing this one as I am still waiting for the go-team to approve the PR, but having some troubles getting people to review it. Also not worth including in release as provides no value.

**Comment by cammellos** [comment]
This can be tested as soon as the builds are ready, but please do not merge before 0.9.30 is cut, no need to include this.

**Comment by chadyj** [comment]
Heya excited to see this taking shape!

I grabbed the iOS and Mac versions from [here](https://github.com/status-im/status-react/pull/6242#issuecomment-430241566).

I hit + on iOS and pair on desktop but I only see 1 device on each. Am I missing something?

**Comment by cammellos** [comment]
@chadyj that's interesting, you should see 2 devices on each if you pressed `+` & `pair` on both. Is the id the same or they are different?

**Review by janherich — APPROVED** [review]

**Comment by chadyj** [comment]
Ping @errorists to take a look at the UI

**Comment by churik** [comment]
### 1. When you restored account that was used before for chatting, no 1-1 chats are restored

On current nightly when you restored an account, all 1-1 chats that were restored as well (https://github.com/status-im/status-react/issues/6417)
For now if you restore account from seed phrase, no 1-1 chats are restored.
This was intentionally removed here or an issue?

### 2. Constant "Fetching messages" on both mobile and desktop
It never disappears, always 1 request left.
1req

### 3. History for public chats is not fetched if user on one device was offline when join to chat on another device (reproducible for both mobile and desktop)

**Prerequisites:** User A uses same account on mobile and desktop, User B on anther (3d) device - no matter desktop or mobile (used current nightly 24/10/2018 on Android)

**Steps:**
- User A, mobile: create new channel, i.e. #testoff 
- User A, mobile: send message "aaa" to  #testoff 
- User A, mobile: log out
- User B: send "bbbb" to #testoff 
- User A, desktop: log in
- User A, desktop: join #testoff and wait when history will be fetched.

**Expected result:** all messages are fetched
**Actual result:** history for channel is empty
**Logs:**
[Status.txt](https://github.com/status-im/status-react/files/2510218/Status.txt)
[geth.log](https://github.com/status-im/status-react/files/2510224/geth.log)

**Comment by cammellos** [comment]
@churik I will look into the issue, but they seem to be related to mailserver connectivity (this PR does not touch that part of code), which mailserver are you currently connected to?

**Comment by cammellos** [comment]
@churik nevermind, I think it might be that current status-go is incompatible, investigating

**Comment by churik** [comment]
@cammellos should I continue testing in this case?

**Comment by cammellos** [comment]
let's wait for the new build, I have reverted status-go to a previous version, I'll test it to make sure it works and let you know

**Comment by cammellos** [comment]
@churik ready to be tested again, latest version of status-go is incompatible with status-react, so I had to rollback, sorry for the inconvenience.

**Comment by churik** [comment]
1 - 3 - fixed.

4. IDs are not clear at all, I even have no info which id from list is mine (if I need to sync contacts I need to know what item should be chosen from the list) 
So for now I suggest at least mark somehow which ID is current, but ideally to see devices in human- readable format.
current

5. Devices are counted regardless of builds, so for example if I have 3 devices, on 2 devices I have builds from https://github.com/status-im/status-react/pull/6242#issuecomment-432640735  installed, on 3 device I have current nightly - I'll see 3 devices instead of 2.
IMO should detect only devices which are able to sync

Still testing, so will update this comment.

**Comment by churik** [comment]
STATUS: WIP (testing)

Tested cases:
1) User A (desktop) adds User B (public chat, no messages in 1-1) -> sync with User A (IOS): sync success, empty chat on IOS appears
2) User B (nightly) adds User A ->  User A (IOS) added a contact when  User A (desktop) is log out -> sync  User A (desktop): sync success
3) User B (release) sends User A  a message -> User A is offline everywhere -> open chat on desktop, send a message; sync (desktop): IOS no contact is added - go offline for IOS again - add contact on desktop  open chat in mobile add contact - no errors; -> sync on both devices - no errors
4) create public chat as User A (IOS) -> send a message, close app; close app on desktop as well -> checked that can join as User B and User A (desktop) -> all messages fetched

**Comment by cammellos** [comment]
@churik 4 I will address probably when UX takes a look at the flow, this PR only adds the mechanics with a bare-bone UI, 5 I could do, but rather do on a separate PR, just not to add too much untested code.
Thanks for reporting.

**Comment by churik** [comment]
@cammellos what about deleting chat? For example if you create a group chat on device A, than delete it - this chat will be preserved on device B and if you send a message to it - it reappears on device A. It is expected right now?

**Comment by cammellos** [comment]
@churik yes, expected, deletions are not synced across devices, also the logic behind deleting a chat is being tested already by @lukaszfryc , but works as such:

1) If you are the admin, when you delete the chat and someone else sends another message, the chat will pop up again (that is until we allow admin to add more admins, otherwise the chat would not have any admin)
2) If you were not the admin (the user who created the chat), when deleting the chat, you will not see it anymore even if new messages are sent

Next step would be to polish this flow and make it clearer for the users.

**Comment by cammellos** [comment]
for reference https://github.com/status-im/status-react/pull/6142

**Comment by churik** [comment]
@cammellos one more question - for now if I have 3 devices (i.e. A, B, C) and tap to device A in "Devices" section on device B -> device A will be synced for both B and C. Is it expected?

**Comment by cammellos** [comment]
@churik correct, currently we sync all devices, pending UX decision

**Comment by SlizkovaGanna** [comment]
@cammellos 
If you were not the admin (the user who created the chat), when deleting the chat, you will not see it anymore even if new messages are sent
For me on mobile, after I delete a chat and the other user responds in it, it re-appears back with a new message.
It's not an issue with current PR most likely.

@cammellos to answer the question below - yes, all of the users were on this PR.

**Comment by cammellos** [comment]
@SlizkovaGanna I will take a look at it, probably not an issue with current PR, but I will check and fix in a separate PR if that's the case.

Just an info, are the members of the group chat all running this PR? (It's possibly incompatible with nightly when it comes to group chats, as never enabled in them, but usable in desktop as it ignores the env settings)

**Comment by churik** [comment]
@cammellos and last questions (or probably issues) from me:
**6. When you restore an account who is member of group chat, group chat is not restored even after sync (can be group-chat issue)**

**7. Devices are not deleted; so if you once installed build on device A, then reinstalled it again - you'll have  two IDs under devices on device B. (expected?)**


**8. When you add to group chat user from current nightly - user cannot see this chat (on desktop and mobile)**

**Steps:**
- install build from https://github.com/status-im/status-react/pull/6242#issuecomment-432947525 on 1 device (User A)
- install nightly build on 2 device (User B)
- as User A create group chat with User B and send a message.

**Expected result:**
user B can see group chat
**Actual result:**
no new chats

**Comment by cammellos** [comment]
6) When you restore an account, you should not see the group chat. To start seeing the group chat it's a bit tricky at the moment, we need to polish the flow, but basically you would have to pair you devices, and then send a message in the group chat from the device that was in the group chat, at that point it should see the group chat (eventually we'll work on synchronizing chats so that as soon as you synchronize it would see the chat).

7) This is expected, we don't remove devices yet, it also a bit of a tricky case (we can't know that a user has uninstalled the app)

8) This is also expected, on mobile group chats are not enabled in nightlies, on desktop they were enabled but that's due to a bug in the build system, they were not meant to be shown so the version might be incompatible.

**Comment by chadyj** [comment]
Great progress here.

It might be nice to introduce device names rather than the current identifier.  Will make it 9999x easier to know what's what with multiple devices.

@errorists hope you have been following along :)

**Comment by cammellos** [comment]
Just a few things that might help (or not) to guide our decisions:

1) They are not really devices, but installations rather (uninstall/install you get a new id, that's because all the data is erased, so having a device-id is to no use to us)
2) We could use random-names, similarly to 3-words, possibly different format etc or user-defined
3) Other info available: type of installation (android,ios,desktop etc), last-seen

**Comment by churik** [comment]
@cammellos one more thing - after 8 installations and 14 contacts I lost ability to sync devices (can repro from all 3 devices - IOS, Android, Mac OSx). Waiting/relogin doesn't help.
Video from desktop + android: http://take.ms/IE1pT
Logs:
[Status.log](https://github.com/status-im/status-react/files/2514789/Status.log)
[geth.log](https://github.com/status-im/status-react/files/2514791/geth.log)

Cannot reproduce it on 3-5 contacts.

<blockquote><div>Monosnap screenshot tool</div><div><strong><a href="https://monosnap.com/file/X9gqkT54EeQAhLRuZjAR0k4x9uf98N">File "screencast 2018-10-25 14-11-29.mp4"</a></strong></div><div>Monosnap — the best tool to take & share your screenshots.</div></blockquote>

**Comment by churik** [comment]
To summarize
I tested:

#### Contact requests
1) User A (desktop) adds User B (public chat, no messages in 1-1) -> sync with User A (IOS): sync success, empty chat on IOS appears
2) User B (nightly) adds User A ->  User A (IOS) added a contact when User A (desktop) is log out -> sync  User A (desktop): sync success
3) User B (release) sends User A  a message -> User A is offline everywhere -> open chat on desktop, send a message; sync (desktop): IOS no contact is added - go offline for IOS again - add contact on desktop  open chat in mobile add contact - no errors; -> sync on both devices - no errors
4) create public chat as User A (IOS) -> send a message, close app; close app on desktop as well -> checked that can join as User B and User A (desktop) -> all messages fetched
5) Sync three divices (Android, IOS, desktop) when all are online, when some of them is log out or offline, when two are offline
6) User B (release) sends User A  a message -> add a contact on IOS and open same chat on desktop -> add contact, sync: no "Add contact" in 1-1 chat anymore
7) https://github.com/status-im/status-react/issues/5204 scenario

#### Group chats
8) Creating group chat - messages sent from IOS, Android, Desktop are visible on all devices
9) Messages sent from IOS when other devices are offline - after login history is fetched
10) Add new contact (R) on IOS -> add them to group chat -> can see that R is a member of group chat -> sync -> R is contact on Android and desktop -> open "Add new member" to group chat -> cannot see R
11) possible group chat members are udated depending on your contact list
12) Messaging in group chat: User A (IOS, Android, Desktop), User B (IOS, Android) - all messages are visible for both User A  and User B

#### Public chats
13) Creating public chat - messages sent from IOS, Android, Desktop are visible on all devices
14) Creating public chat on current nightly, send several messages - history can be fetched on three devices.

#### General
15) /send /request command in 1-1 chats
16) deleting/clearing history is not synced
17) chats can be synced regardless of mailservers and network (IOS can use mailserver 1 and Ropsten, Android - mailserver 2 and Mainnet)
18) Opened DApps \ links are not synced across devices

#### Questions

1. IDs are not clear and counts all installtions of any builds (should create separate issue?) 
2. Group chats are reappearing after deleting if new message is coming (regadless of are you admin ar not) and sctually deleting are not synced (should create separate issue?)
3. Group chats after restoring account - separate issue?
4. https://github.com/status-im/status-react/pull/6242#issuecomment-433028644 - should be fixed.


Please tell me if I missed smth.

**Comment by cammellos** [comment]
@churik ready to be tested again, it should now sync.

I  though  had to make a change, seems like some pictures are too large to be sent this way, so I have disabled syncing of the picture (all your contact will sync, but the default picture will be on the other device, unless you already had that contact's picture).
I will create an issue to address this separately. 1, 2 and 3 they will be separate issues, correct.
Thanks!

**Comment by churik** [comment]
@cammellos retested last change, it works for me, so all theree devices are synced contacts, on one of them.
One more small issue appeared - if you sync contacts, some of them as you said are shown with default picture. So if you tap on it from the list to start new chat - you will see default picture for public chat, not defeult for 1-1.
Video: http://take.ms/u5ogp
Screemshot:
jjjj
Can it be fixed here or I should create separate issue, WDYT?

**Comment by cammellos** [comment]
@churik Thanks again for taking the time, I would create a separate issue if you don't mind, as I am keen to get this in, but I will push out a PR soon, it should be easy to fix.

**Comment by churik** [comment]
@cammellos 
5) for new users there is no option to "Start Group Chat" on mobile. Is it intentionally?
I used build https://github.com/status-im/status-react/pull/6242#issuecomment-433203146. 
6) after last changes messages between group chats are not synced; some of them are lost on one of three devices.
I couldn't find reproducible steps, but there is log when I sent 5 messages on IOS, 2 of them are lost on Desktop. The same situation I got with IOS
[geth.log](https://github.com/status-im/status-react/files/2521712/geth.log)

**Comment by cammellos** [comment]
@churik you need to enable development mode to see the option, thanks for
the logs i will investigate

**Comment by churik** [comment]
@cammellos got one reproducible way for lost messages:
on device A you should send a message to group chat when device B is offline or app is closed, so when you get back to group chat on device B message will be never fetched.
Log: [geth.log](https://github.com/status-im/status-react/files/2521758/geth.log)

**Comment by churik** [comment]
#### Summary:

Known issues (will be created separately):
1. IDs are not clear and counts all installtions of any builds 
2. Group chats are reappearing after deleting if new message is coming (regardless of are you admin or not) and actually deleting are not synced 
3. Group chats after restoring account
~~4. Messages for group chats are not fetched for device B if they were sent from device A when device B was offline (for the same account)~~
5.  If you sync contacts, some of them as you said are shown with default picture. So if you tap on it from the list to start new chat - you will see default picture for public chat, not default for 1-1.

**Group chats and pairing should be off in nightly**

**Pull request merged**